### ISSUE TITLE
How to initialize ring_buffer

### ISSUE THREAD

**Issue by kgreenek** [RustAudio/sample#94]
I want to use the sample library to create a class for calculating the auto-correlation function of a signal. In principle, the implementation should be fairly similar to rms.rs. However, I'm having some trouble creating a ring_buffer.

I'm definitely a rust beginner, so I am certain I'm just doing something wrong!

Here is the best I've come up with so far:
```rust
extern crate sample;

use sample::ring_buffer::Slice;
use std::marker::PhantomData;

#[derive(Clone)]
struct DiffSquaredAcf<F, S> where F: sample::Frame, S: sample::ring_buffer::Slice<Element = F::Float>
{
    frame: PhantomData<F>,
    frame_buffer: sample::ring_buffer::Fixed<S>,
    window_size: usize,
    max_t: usize,
}
 
impl<F, S> DiffSquaredAcf<F, S>
where
    F: sample::Frame,
    S: sample::ring_buffer::Slice<Element = F::Float>,
{
    pub fn new(window_size: usize, max_t: usize) -> Self {
        let buffer_size = window_size + max_t;
        let frame_vec = vec![F::equilibrium().to_float_frame(); buffer_size];
        let mut frame_buffer = sample::ring_buffer::Fixed::from(frame_vec);
        DiffSquaredAcf {
            frame: PhantomData,
            // This line fails with a type error
            frame_buffer: frame_buffer,
            window_size: window_size,
            max_t: max_t,
        }
    }

    pub fn add_frame(&mut self, frame: F) {
        //...
    }
}
```

I think I'm close, but I just can't figure out how to create the ring_buffer in a generic way. In the above example, I get the following error:
```bash
error[E0308]: mismatched types
  --> /Users/kevin/Projects/kgreenek/rzero/src/diff_squared_acf.rs:26:27
   |
26 |             frame_buffer: frame_buffer,
   |                           ^^^^^^^^^^^^ expected type parameter, found struct `std::vec::Vec`
   |
   = note: expected type `sample::ring_buffer::Fixed<S>`
              found type `sample::ring_buffer::Fixed<std::vec::Vec<<F as sample::Frame>::Float>>`

error: aborting due to previous error

error: Could not compile `rzero`.

To learn more, run the command again with --verbose.
```

I would think this should work because there is an impl for Vec for the ring_buffer::Slice trait. But it doesn't :(

It's similar to the Rms class, but in my case, I don't want to have my ring_buffer passed to the constructor as an argument, I want to create it in the new method. The DiffSquaredAcf struct will obviously have more members; right now I'm just trying to figure out how to create this internal ring_buffer!

Also, I'd love to contribute this upstream when it's ready, if it's something you'd be interested in. The sample module is great!

**Comment by mitchmindtree** [comment]
Hey thanks for the issue and the kind words!

The reason you're getting an error is that your implementation suggests that you return a `DiffSquaredAcf<F, S>`, when what you're actually returning is a `DiffSquaredAcf<F, Vec<F as sample::Frame>::Float>>`.

Seeing as you're using a `Vec` for the frame buffer in `DiffSquaredAcf::new`, it looks like you might want to either:

1. Change the `frame_buffer` field type in the `DiffSquaredAcf` struct to `Vec<<F as sample::Frame>::Float>` or
2. Implement the `new`  constructor for `DiffSquaredAcf<F, Vec<<F as sample::Frame>::Float>>` (instead of `DiffSquaredAcf<F, S>` which suggests a generic type).

Alternatively, if you want your constructor to remain generic over its `ring_buffer`, consider taking the `ring_buffer` as an argument and taking the window_size from that, rather than trying to construct one from a given `window_size`.

Hope this helps!

**Comment by kgreenek** [comment]
This was very helpful, thanks! I think I'm finally getting the hang of Rust's generics; coming from C++.

I'm implementing the YIN algorithm for pitch detection in Rust, and it seemed to make sense to push the auto-correlation class upstream. See: https://github.com/RustAudio/sample/pull/95.

In my case, I needed a modified "diff squared" correlation function, rather than the standard acf where you multiply the values and sum the result together. There are few common minor variations of the acf that are used, so I made a generic Acf trait in case someone else wanted to implement something slightly different.

**Comment by andrewcsmith** [comment]
Hi @kgreenek, I contribute to this lib fairly regularly and thought you may also be interested in my [vox_box](https://github.com/andrewcsmith/vox_box.rs) library. I just published 0.3, incorporating some of my latest changes. I'd greatly welcome a PR implementing the `Autocorrelate` trait for `T: ring_buffer::Slice`. Here's an issue if you're interested: https://github.com/andrewcsmith/vox_box.rs/issues/10

**Comment by kgreenek** [comment]
vox_box looks sweet! Thanks for sharing that.

For my project, I am doing real-time processing. So I need the ability to add one sample, calculate the ACF, add one sample, calculate the ACF, etc. If you look at #95, I am able to get a huge optimization by designing it incrementally like that. I only have to add the new ACF value for each sample, and subtract the old one (rather than having to multiply and sum over the whole window for each offset, every time a new sample is added).
